### PR TITLE
Job API enabler [v3]

### PR DESCRIPTION
**Pull request by clebergnu**
Set of changes on the way to make the Job API a public feature.  The main changes include:

 * The conversion from an `argparse.Namespace` based instance to configure most job aspects, to a more intuitive dictionary
 * The rename of `args` to `config` as the attribute name in `Job` to hold that configuration
 * The automatic configuration of the output (if not done by the app)
 * Some example jobs

---

Changes from v2 (#3234):
 * Compliance with W0221 checks
 * Rebased and resolved conflicts

Changes from v1 (#3219):
 * Merged the rename transformation from the `argparse.Namespace` based args to dict with the rename from `args` to `config` as they touched basically the same code twice
 * Removed commit `Job: allow a job to define multiple test suites` as that wasn't being used here

### CODE REVIEW

File: avocado/core/job.py [hunk 1]
```diff
@@ -137,16 +137,16 @@ def __init__(self, args=None):
         #: only once, since they are read only and will be shared across all
         #: tests of a job.
         self.test_parameters = None
-        if "test_parameters" in self.args:
+        if "test_parameters" in self.config:
             self.test_parameters = {}
-            for parameter_name, parameter_value in self.args.test_parameters:
+            for parameter_name, parameter_value in self.config.get('test_parameters'):
```
Comment (wainersm) [inline avocado/core/job.py]: I think it needs to iterate over `items()`.

File: avocado/core/job.py [hunk 2]
```diff
@@ -336,8 +335,8 @@ def soft_abort(msg):
                 os.unlink(proc_latest)
 
     def _start_sysinfo(self):
-        if hasattr(self.args, 'sysinfo'):
-            if self.args.sysinfo == 'on':
+        if 'sysinfo' in self.config:
+            if self.config.get('sysinfo') == 'on':
```
Comment (wainersm) [inline avocado/core/job.py]: I would combine those ifs in one ```if self.config.get('sysinfo', None) == 'on'```

File: avocado/core/job.py (hunk 1 above)
Comment (clebergnu) [reply avocado/core/job.py]: This is actually a list of tupples (see avocado/plugin/run.py:Run._test_parameters).

File: avocado/core/job.py (hunk 2 above)
Comment (clebergnu) [reply avocado/core/job.py]: Yep, I tried to do a more straightforward conversion to ease review, but I was bothered by this also.  I'll add a commit changing the style all over on top of those.

File: optional_plugins/result_upload/avocado_result_upload/__init__.py [hunk 3]
```diff
@@ -42,12 +42,12 @@ def render(self, result, job):
...
         self.upload_url = None
-        if getattr(job.args, 'result_upload_url', None) is not None:
-            self.upload_url = job.args.result_upload_url
+        if job.config.get('result_upload_url', None) is not None:
+            self.upload_url = job.config.get('result_upload_url')
 
         self.upload_cmd = None
-        if getattr(job.args, 'result_upload_cmd', None) is not None:
-            self.upload_cmd = job.args.result_upload_cmd
+        if getattr(job.config, 'result_upload_cmd', None) is not None:
```
Comment (wainersm) [inline optional_plugins/result_upload/avocado_result_upload/__init__.py]: ```if job.config.get('result_upload_cmd', None) is not None```

File: optional_plugins/runner_docker/avocado_runner_docker/__init__.py [hunk 4]
```diff
@@ -120,11 +120,11 @@ def __init__(self, job, result):
         self.remote = None      # Will be set in `setup`
 
     def setup(self):
-        dkrcmd = self.job.args.docker_cmd
-        dkr_opt = self.job.args.docker_options
+        dkrcmd = self.job.args.get('docker_cmd')
```
Comment (wainersm) [inline optional_plugins/runner_docker/avocado_runner_docker/__init__.py]: At this point the ```job``` object hasn't ```args``` (rather ```config```), it has?

File: optional_plugins/result_upload/avocado_result_upload/__init__.py (hunk 3 above)
Comment (clebergnu) [reply optional_plugins/result_upload/avocado_result_upload/__init__.py]: Yep, nice catch!

File: optional_plugins/runner_docker/avocado_runner_docker/__init__.py (hunk 4 above)
Comment (clebergnu) [reply optional_plugins/runner_docker/avocado_runner_docker/__init__.py]: Yes, it should have `.config` at this point.  Nice catch.

File: avocado/core/parser.py [hunk 5]
```diff
@@ -138,11 +138,13 @@ def finish(self):
 
         Side effect: set the final value for attribute `args`.
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/parser.py]: Yep!

File: avocado/core/job.py (hunk 2 above)
Comment (clebergnu) [reply avocado/core/job.py]: Actually, I thought there were many more like this, but I only found one.  So, I'm putting them on the same (first) commit.

File: avocado/core/parser.py [hunk 6]
```diff
@@ -138,11 +138,13 @@ def finish(self):
...
+        args, extra = self.application.parse_known_args(namespace=self.args)
         if extra:
             msg = 'unrecognized arguments: %s' % ' '.join(extra)
             for sub in self.application._subparsers._actions:  # pylint: disable=W0212
                 if sub.dest == 'subcommand':
                     sub.choices[self.args.subcommand].error(msg)
 
             self.application.error(msg)
+        # from this point on, config is a dictionary based on a argparse.Namespace
+        self.config = vars(args)
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/parser.py]: Yep, good point.

**Pull request closed without merging**